Add Function max_intersections: calc the maximum number of overlapping intervals

ArNine opened the pull request:
hi duckdb team:
I've found that many olap have similar functionality, but duckdb doesn't.Therefore, I would like to try my best to contribute something small to duckdb.
This pr add an aggregate function `max_intersections` that calculates the maximum number of intersecting intervals using a scanline algorithm. It can solve problems such as the maximum number of concurrent users in a game or a meeting.

It yields the same result as the SQL statement below, but is more concise. 

```
create table (start_time int, end_time int);

select
    max(current_count)
from
(
    select
        sum(flag) over(order by time_point, change desc) as current_count
    from
    (
        select start_time as time_point, 1 as flag
        from table
        union all
        select end_time as time_point, -1 as flag
        from table
    )
);

VS

select
max_intersections(start_time, end_time)
from table;
```

ArNine committed:
add function max_intersections

ArNine committed:
format code

hawkfish commented on extension/core_functions/include/core_functions/aggregate/distributive_functions.hpp:
```diff
@@ -299,4 +299,14 @@ struct SumNoOverflowFun {
 	static AggregateFunctionSet GetFunctions();
 };
 
+struct MaxIntersectionsFun {
```
This is a holistic aggregate (it requires all the data) so it should be in the holistic section.

hawkfish commented on test/sql/aggregate/aggregates/test_max_intersections.test:
```diff
@@ -0,0 +1,107 @@
+# name: test/sql/aggregate/aggregates/test_max_intersections.test
```
Can we add tests for windowing (OVER clause)? I suspect you will can't a custom window function.

hawkfish commented on extension/core_functions/aggregate/distributive/max_intersections.cpp:
```diff
@@ -0,0 +1,203 @@
+			return;
+		}
+
+		target.starts.reserve(target.starts.size() + source.starts.size());
```
This will copy the data - which is great for windowing, but if you can move the data when `AggregateInputData .combine_type == AggregateCombineType::ALLOW_DESTRUCTIVE` you will reduce the memory footprint during normal aggregation.

hawkfish reviewed (changes requested):
The windowing support and testing is the big thing that is missing.

ArNine committed:
Adjust function max_intersections code structure, optimize max_intersection performance, and add test cases

ArNine commented:
> The windowing support and testing is the big thing that is missing.

hi, very thanks review, i have completed the revisions, and last happy new year for you

lnkuiper commented on extension/core_functions/aggregate/holistic/max_intersections.cpp:
```diff
@@ -0,0 +1,210 @@
+		// Use parallel sorting when dataset is large enough to benefit from threading
+		const size_t PARALLEL_SORT_THRESHOLD = 10000;
+
+		if (num_intervals >= PARALLEL_SORT_THRESHOLD) {
```
Please remove the multi-threaded implementation and keep only the single-threaded implementation. DuckDB has it's own threading model, and should never spawn threads within any scalar/aggregate function

lnkuiper commented on extension/core_functions/aggregate/holistic/max_intersections.cpp:
```diff
@@ -0,0 +1,210 @@
+
+} // namespace
+
+AggregateFunction MaxIntersectionsFun::GetFunction() {
```
This fails to build on Windows:
```
D:\a\duckdb\duckdb\extension\core_functions\aggregate\holistic\max_intersections.cpp(196,19): error C2653: 'MaxIntersectionsFun': is not a class or namespace name [D:\a\duckdb\duckdb\extension\core_functions\aggregate\holistic\duckdb_core_functions_holistic.vcxproj]
```
Looks like it might be a missing include.

lnkuiper reviewed (changes requested):
Thanks for the PR! This function seems very useful :)

I've built this PR and tried it out a bit, but I am sceptical about correctness and performance. I ran some queries on the `lineitem` table from TPC-H at scale factor 10 (generated with `CALL dbgen(sf=10);`.

You suggested that this SQL:
```sql
select
    max(current_count)
from
(
    select
        sum(flag) over(order by time_point, flag desc) as current_count
    from
    (
        select l_commitdate as time_point, 1 as flag
        from lineitem
        union all
        select l_receiptdate as time_point, -1 as flag
        from lineitem
    )
);
```
(I've had to infer that it needed to be ordered by `flag desc`, as the column `change` in your PR description does not exist)

Should be equivalent to:
```sql
select max_intersections(epoch(l_commitdate)::bigint, epoch(l_receiptdate)::bigint) from lineitem;
```

But this generates different results (`476722` vs `669632`). Performance does seem OK, the specialized function takes 1.6s whereas the window function takes 2.5s.

However, when we add groups, this story changes. I've grouped this by `l_suppkey` (100K groups) and `l_partkey` (2M groups). Queries:
```sql
select
    l_suppkey, max(current_count)
from
(
    select
        l_suppkey, sum(flag) over(partition by l_suppkey order by time_point, flag desc) as current_count
    from
    (
        select l_suppkey, l_commitdate as time_point, 1 as flag
        from lineitem
        union all
        select l_suppkey, l_receiptdate as time_point, -1 as flag
        from lineitem
    )
)
group by l_suppkey;

select l_suppkey, max_intersections(epoch(l_commitdate)::bigint, epoch(l_receiptdate)::bigint) from lineitem group by l_suppkey;

select
    l_partkey, max(current_count)
from
(
    select
        l_partkey, sum(flag) over(partition by l_partkey order by time_point, flag desc) as current_count
    from
    (
        select l_partkey, l_commitdate as time_point, 1 as flag
        from lineitem
        union all
        select l_partkey, l_receiptdate as time_point, -1 as flag
        from lineitem
    )
)
group by l_partkey;

select l_partkey, max_intersections(epoch(l_commitdate)::bigint, epoch(l_receiptdate)::bigint) from lineitem group by l_partkey;
```

When grouped by `l_suppkey`, the Window function implementation is faster (5.2s vs.  8.0s), and when grouped by `l_partkey`, the the window function remains fast (3.7s), but the specialized implementation slows down drastically (to a whopping 42.8s!).

This probably has something to do with the window function sorting large batches of data, even when there are many groups, while the specialized function has to sort many small batches, or due to allocating millions of small `std::vector`, which can bottleneck performance.

ArNine committed:
max_intersections remove mult-thread, reduce the number of vector, add function overload

ArNine commented:
> Thanks for the PR! This function seems very useful :)
> 
> I've built this PR and tried it out a bit, but I am sceptical about correctness and performance. I ran some queries on the `lineitem` table from TPC-H at scale factor 10 (generated with `CALL dbgen(sf=10);`.
> 
> You suggested that this SQL:
> 
> ```sql
> select
>     max(current_count)
> from
> (
>     select
>         sum(flag) over(order by time_point, flag desc) as current_count
>     from
>     (
>         select l_commitdate as time_point, 1 as flag
>         from lineitem
>         union all
>         select l_receiptdate as time_point, -1 as flag
>         from lineitem
>     )
> );
> ```
> 
> (I've had to infer that it needed to be ordered by `flag desc`, as the column `change` in your PR description does not exist)
> 
> Should be equivalent to:
> 
> ```sql
> select max_intersections(epoch(l_commitdate)::bigint, epoch(l_receiptdate)::bigint) from lineitem;
> ```
> 
> But this generates different results (`476722` vs `669632`). Performance does seem OK, the specialized function takes 1.6s whereas the window function takes 2.5s.
> 
> However, when we add groups, this story changes. I've grouped this by `l_suppkey` (100K groups) and `l_partkey` (2M groups). Queries:
> 
> ```sql
> select
>     l_suppkey, max(current_count)
> from
> (
>     select
>         l_suppkey, sum(flag) over(partition by l_suppkey order by time_point, flag desc) as current_count
>     from
>     (
>         select l_suppkey, l_commitdate as time_point, 1 as flag
>         from lineitem
>         union all
>         select l_suppkey, l_receiptdate as time_point, -1 as flag
>         from lineitem
>     )
> )
> group by l_suppkey;
> 
> select l_suppkey, max_intersections(epoch(l_commitdate)::bigint, epoch(l_receiptdate)::bigint) from lineitem group by l_suppkey;
> 
> select
>     l_partkey, max(current_count)
> from
> (
>     select
>         l_partkey, sum(flag) over(partition by l_partkey order by time_point, flag desc) as current_count
>     from
>     (
>         select l_partkey, l_commitdate as time_point, 1 as flag
>         from lineitem
>         union all
>         select l_partkey, l_receiptdate as time_point, -1 as flag
>         from lineitem
>     )
> )
> group by l_partkey;
> 
> select l_partkey, max_intersections(epoch(l_commitdate)::bigint, epoch(l_receiptdate)::bigint) from lineitem group by l_partkey;
> ```
> 
> When grouped by `l_suppkey`, the Window function implementation is faster (5.2s vs. 8.0s), and when grouped by `l_partkey`, the the window function remains fast (3.7s), but the specialized implementation slows down drastically (to a whopping 42.8s!).
> 
> This probably has something to do with the window function sorting large batches of data, even when there are many groups, while the specialized function has to sort many small batches, or due to allocating millions of small `std::vector`, which can bottleneck performance.

hi thanks for review:) 
I reproduced your test. The different results (476722 vs 669632) are because this function filters out cases where `start > end`, therefore the SQL should be
```
select
    max(current_count)
from
(
    select
        sum(flag) over(order by time_point, flag desc) as current_count
    from
    (
        select l_commitdate as time_point, 1 as flag
        from lineitem
        where l_commitdate <= l_receiptdate 
        union all
        select l_receiptdate as time_point, -1 as flag
        from lineitem
        where l_commitdate <= l_receiptdate 
    )
);
```

I've removed multi-threaded, added function overload, and reduced the number of vectors, but it still takes about 20 seconds when performing a `group by l_partkey`. I will try to improve performance again.

ArNine committed:
Fix max_intersections  flag bit error
